flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/source/enumerator/splitter/MongoSampleSplitter.java

```diff
@@ -0,0 +1,131 @@
+import com.mongodb.client.model.Projections;
+import com.mongodb.client.model.Sorts;
+import org.bson.BsonDocument;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+import static org.apache.flink.connector.mongodb.common.utils.MongoConstants.BSON_MAX_KEY;
+import static org.apache.flink.connector.mongodb.common.utils.MongoConstants.BSON_MIN_KEY;
+import static org.apache.flink.connector.mongodb.common.utils.MongoConstants.ID_FIELD;
+import static org.apache.flink.connector.mongodb.common.utils.MongoConstants.ID_HINT;
+
+/**
+ * Sample Partitioner
+ *
+ * <p>Samples the collection to generate partitions.
+ *
+ * <p>Uses the average document size to split the collection into average sized chunks
+ *
+ * <p>The partitioner samples the collection, projects and sorts by the partition fields. Then uses
+ * every {@code samplesPerPartition} as the value to use to calculate the partition boundaries.
+ *
+ * <ul>
+ *   <li>scan.partition.size: The average size (MB) for each partition. Note: Uses the average
+ *       document size to determine the number of documents per partition so may not be even.
+ *       Defaults to: 64mb.
+ *   <li>scan.partition.samples: The number of samples to take per partition. Defaults to: 10. The
+ *       total number of samples taken is calculated as: {@code samples per partition * (count of
+ *       documents / number of documents per partition)}.
+ * </ul>
+ */
+@Internal
+public class MongoSampleSplitter implements MongoSplitters.MongoSplitter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoSampleSplitter.class);
+
+    public static final MongoSampleSplitter INSTANCE = new MongoSampleSplitter();
+
+    private MongoSampleSplitter() {}
+
+    @Override
+    public Collection<MongoScanSourceSplit> split(MongoSplitContext splitContext) {
+        MongoReadOptions readOptions = splitContext.getReadOptions();
+        MongoNamespace namespace = splitContext.getMongoNamespace();
+
+        long count = splitContext.getCount();
+        long partitionSizeInBytes = readOptions.getPartitionSize().getBytes();
+        int samplesPerPartition = readOptions.getSamplesPerPartition();
+
+        long avgObjSizeInBytes = splitContext.getAvgObjSize();
+        long numDocumentsPerPartition = partitionSizeInBytes / avgObjSizeInBytes;
+
+        if (numDocumentsPerPartition >= count) {
+            LOG.info(
+                    "Fewer documents ({}) than the number of documents per partition ({}), fallback a SingleSplitter.",
+                    count,
+                    numDocumentsPerPartition);
+            return MongoSingleSplitter.INSTANCE.split(splitContext);
+        }
+
+        int numberOfSamples =
+                (int) Math.ceil((samplesPerPartition * count * 1.0d) / numDocumentsPerPartition);
+
+        List<BsonDocument> samples =
```

Comment:
I find it very confusing that this is named "samples"; I would expect this to be used to derive _some_ quality of the whole dataset (like `avgObjSizeInBytes`), but these are rather pointers into the key space / partition boundaries.